consensus/types/src/beacon_state.rs

```diff
@@ -1354,24 +1378,110 @@ impl<T: EthSpec> BeaconState<T> {
         ))
     }
 
-    /// Return the activation churn limit for the current epoch (number of validators who can enter per epoch).
-    ///
-    /// Uses the epoch cache, and will error if it isn't initialized.
-    ///
-    /// Spec v1.4.0
-    pub fn get_activation_churn_limit(&self, spec: &ChainSpec) -> Result<u64, Error> {
-        Ok(match self {
+    /// TODO: consider removing fork guard?
+    pub fn switch_to_compounding_validator(
+        &mut self,
+        validator_index: usize,
+        spec: &ChainSpec,
+    ) -> Result<(), Error> {
+        match self {
             BeaconState::Base(_)
             | BeaconState::Altair(_)
             | BeaconState::Merge(_)
-            | BeaconState::Capella(_) => self.get_churn_limit(spec)?,
-            BeaconState::Deneb(_) | BeaconState::Electra(_) => std::cmp::min(
-                spec.max_per_epoch_activation_churn_limit,
-                self.get_churn_limit(spec)?,
-            ),
+            | BeaconState::Capella(_)
+            | BeaconState::Deneb(_) => Err(Error::IncorrectStateVariant),
+            BeaconState::Electra(_) => {
+                let validator = self.get_validator_mut(validator_index)?;
+                if validator.has_eth1_withdrawal_credential(spec) {
+                    validator.set_compounding_withdrawal_credentials(spec);
+                    self.queue_excess_active_balance(validator_index, spec)?;
+                }
+                Ok(())
+            }
+        }
+    }
+
+    /// TODO: consider removing fork guard?
```

Comment:
Instead of matching on the fork you should match on `self.pending_balance_deposits_mut()` returning Ok()